primitives/src/lib.rs

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! # Rust Bitcoin - primitive types.
+//!
+//! Primitive data types that are used throughout the [`rust-bitcoin`] ecosystem.
+//!
+//! [`rust-bitcoin`]: <https://github.com/rust-bitcoin>
+
+#![cfg_attr(all(not(test), not(feature = "std")), no_std)]
+// Experimental features we need.
+#![cfg_attr(docsrs, feature(doc_auto_cfg))]
+#![cfg_attr(bench, feature(test))]
+// Coding conventions.
+#![warn(missing_docs)]
+// Exclude lints we don't think are valuable.
+#![allow(clippy::needless_question_mark)] // https://github.com/rust-bitcoin/rust-bitcoin/pull/2134
+#![allow(clippy::manual_range_contains)] // More readable than clippy's format.
+#![allow(clippy::needless_borrows_for_generic_args)] // https://github.com/rust-lang/rust-clippy/issues/12454
+
+#[cfg(bench)]
+extern crate test;
+
+extern crate alloc;
+
+#[cfg(feature = "std")]
+extern crate std;
+
+/// Re-export the `bitcoinconsensus` crate.
+#[cfg(feature = "bitcoinconsensus")]
+pub extern crate bitcoinconsensus;
+
+/// Re-export the `bech32` crate.
+#[cfg(feature = "bech32")]
+pub extern crate bech32;
+
+/// Re-export the `base58ck` crate.
+#[cfg(feature = "base58")]
+pub extern crate base58;
+
+/// Rust implementation of cryptographic hash function algorithms.
+pub extern crate hashes;
```

Comment:
Should not be `pub` - see above

Reply:
I like the aim of not exposing `hashes` in the public API at all. We have to be careful that we remember not to do a breaking release of `hashes` which changes what the newtype macros output then think that because hashes is not exposed that we don't need a breaking release in `bitcoin`.

Reply:
We can have a special, stable macro for that. If anyone tries to touch it it's instant NACK. :)

Reply:
We should have a bot that nacks stuff, that would be cool.